### PR TITLE
SWA

### PR DESCRIPTION
**Pull request by izmailovpavel**
In this pull request I add SWA optimizer to `torchcontrib.optim`. The structure of the pull request is as follows:
- `torchcontrib/optim/swa.py` contains the implementation of `SWA` optimizer class and `bn_update` utility for updating batch normalization statistics for a given model and data loader
- `test/test_swa.py` contains unit tests for `SWA` and `bn_update`; I included (a slightly modified version of) `test_optim` unit tests as well as new tests specific for SWA; I also included `common.py` from `torch/test` to be able to import from it
- `examples/swa/` contain example usage of the implemented optimizer for training deep networks; the code there is adapted from (https://github.com/timgaripov/swa); I tested it with VGG and WRN to make sure that I can replicate the results from our original implementation using the new code; I don't expect `examples/swa` to be merged into `torchcontrib` and I include them to demonstrate the interface of the code; `examples/swa/train.py` shows basic usage of `SWA` class (SWA with constant learning rate) and `examples/swa/train_fastswa.py` shows how to implement a custom cyclical learning rate schedule with SWA

The `SWA` code can be used in two modes: automatic and manual. Automatic mode requires minimal actions from the user. The interface is as follows
```
base_opt = torch.optim.SGD(model.parameters(), lr=0.1)
# create the SWA wrapper of base_opt
opt = torchcontrib.optim.SWA(base_opt, swa_start=10, swa_freq=5, swa_lr=0.05)
for _ in range(100):
     opt.zero_grad()
     loss_fn(model(input), target).backward()
     opt.step()
# Swap the values of the optimized variables and SWA buffers
opt.swap_swa_sgd()
```
Note the `opt.swap_swa_sgd()` method: in the end of training we reset the values of the optimized parameters to the SWA running averages. We do this by swapping the values of the optimizer `swa_buffer` buffers and optimized variables. Then, if we need to continue training (e.g. we evaluated SWA and want to continue training as normal), we can call `opt.swap_swa_sgd()` again and continue training:
```
base_opt = torch.optim.SGD(model.parameters(), lr=0.1)
opt = torchcontrib.optim.SWA(base_opt, swa_start=10, swa_freq=5, swa_lr=0.05)
for _ in range(100):
     opt.zero_grad()
     loss_fn(model(input), target).backward()
     opt.step()
opt.swap_swa_sgd()
<evaluate SWA solution>
opt.swap_swa_sgd()
for _ in range(100):
     opt.zero_grad()
     loss_fn(model(input), target).backward()
     opt.step()
opt.swap_swa_sgd()
```

In the manual mode the user can implement custom learning rate and update schedules for SWA (e.g. update SWA running averages for different parameter groups with different frequency). The interface for the manual mode is as follows:
```
opt = torchcontrib.optim.SWA(base_opt)
for i in range(100):
     opt.zero_grad()
     loss_fn(model(input), target).backward()
     opt.step()
     if i > 10 and i % 5 == 0:
         opt.update_swa()
         # or opt.update_swa_group(opt.param_groups[1]) to update a single param_group
opt.swap_swa_sgd()
```

Auto mode is selected when both `swa_start` and `swa_freq` arguments are provided, and otherwise manual mode is used. `swa_start` is the number of optimization steps before we start SWA averaging, and `swa_freq` is the number of steps between subsequent updates of SWA-averages. In case `swa_lr` is provided, the learning rate of the inner optimizer is reset to `swa_lr` at every step starting from `swa_start` step.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,27 +1,5 @@
...
-    description='implementations of ideas from recent papers',
-    long_description=readme,
-    license='BSD',
-
-    # Package info
-    packages=find_packages(exclude=('test',)),
-
-    zip_safe=True,
+setup(name="PTContrib",
+      packages=["."],
```
Comment (apaszke) [inline setup.py]: Please don't change this file

File: test/test_swa.py [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+    def _build_params_dict(self, weight, bias, **kwargs):
+        return [dict(params=[weight]), dict(params=[bias], **kwargs)]
+
+    def _build_params_dict_single(self, weight, bias, **kwargs):
+        return [dict(params=bias, **kwargs)]
+
+    #Test SWA
+    
+    def test_swa(self):
+        print("here")
```
Comment (apaszke) [inline test/test_swa.py]: This print should be removed

File: test/test_swa.py [hunk 3]
```diff
@@ -0,0 +1,692 @@
+import functools
```
Comment (apaszke) [inline test/test_swa.py]: This file should stay as `test_optim.py`

File: test/test_swa.py [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+            return contriboptim.SWA(
+                lbfgs, swa_start=1000, swa_freq=1, swa_lr=1e-3)
+
+        auto_constructor_list = [sgd_constructor, sgd_momentum_constructor, 
+                                 adam_constructor, adadelta_constructor, 
+                                 adagrad_constructor, adamax_constructor, 
+                                 rmsprop_constructor, rprop_constructor,
+                                 asgd_constructor, lbfgs_constructor]
+
+        #auto_constructor_list = [sgd_momentum_constructor]
```
Comment (apaszke) [inline test/test_swa.py]: Dead code

File: test/test_swa.py [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+
+        for i, constructor in enumerate(auto_constructor_list):
+            print(constructor)
+            # Pass
```
Comment (apaszke) [inline test/test_swa.py]: Unnecessary print and comment

File: test/test_swa.py [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+        y = Variable(torch.Tensor([3., 7.]), requires_grad=True)
+
+        def loss_fun(a, b):
+            return torch.sum(a * b)**2
+
+        opt = optim.SGD([{'params': [x]},
+                        {'params': [y], 'lr': 1e-3}], lr=1e-2, momentum=0.9)
+        return x, y, loss_fun, opt
+
+    def testSWAAuto(self):
```
Comment (apaszke) [inline test/test_swa.py]: Please use `snake_case` just like in the rest of the code base

File: torchcontrib/optim/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+from .swa import SWA, bn_update
```
Comment (apaszke) [inline torchcontrib/optim/__init__.py]: Please make `bn_update` a static member of `SWA` (so someone can do `SWA.bn_update(...)`). Otherwise it pollutes the optim namespace, and its name doesn't indicate that it's really specific to this optimizer

File: test/test_swa.py [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+        n_avg = 0
+        for i in range(1, 11):
+            opt.zero_grad()
+            loss = loss_fun(x, y)
+            loss.backward()
+            opt.step()
+            if i % swa_freq == 0 and i > swa_start:
+                opt.update_swa_group(opt.param_groups[1])
+                n_avg += 1
+                y_sum += y.data
```
Comment (apaszke) [inline test/test_swa.py]: This code keeps repeating over and over in those tests. Is there really no way to factor out this training loop into a separate function? You can pass the updating line as a lambda to be called inside.

File: test/common.py [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+        return path
+    try:
+        data = request.urlopen(url, timeout=15).read()
+        with open(path, 'wb' if binary else 'w') as f:
+            f.write(data)
+        return path
+    except error.URLError:
+        msg = "could not download test file '{}'".format(url)
+        warnings.warn(msg, RuntimeWarning)
+        raise unittest.SkipTest(msg)
```
Comment (apaszke) [inline test/common.py]: What code from this file do you need?

File: test/test_swa.py (hunk 3 above)
Comment (izmailovpavel) [reply test/test_swa.py]: Ok, but note that some of the common code in the `test_optim.py` was modified specifically for the SWA syntax; see e.g. https://github.com/izmailovpavel/contrib/blob/master/test/test_swa.py#L60-L61. Is it ok?

File: test/common.py (hunk 9 above)
Comment (izmailovpavel) [reply test/common.py]: I need `TestCase` and `run_tests`.

File: test/test_swa.py (hunk 3 above)
Comment (apaszke) [reply test/test_swa.py]: Uh ok, it's not perfect, but we can leave it then 😕

File: setup.py (hunk 1 above)
Comment (izmailovpavel) [reply setup.py]: done, reverted it to its original state

File: test/test_swa.py (hunk 2 above)
Comment (izmailovpavel) [reply test/test_swa.py]: done

File: test/test_swa.py (hunk 4 above)
Comment (izmailovpavel) [reply test/test_swa.py]: done

Comment (izmailovpavel) [reply test/test_swa.py]: done

File: test/test_swa.py (hunk 5 above)
Comment (izmailovpavel) [reply test/test_swa.py]: done

File: test/test_swa.py (hunk 6 above)
Comment (izmailovpavel) [reply test/test_swa.py]: done

File: test/test_swa.py (hunk 8 above)
Comment (izmailovpavel) [reply test/test_swa.py]: Done, see https://github.com/izmailovpavel/contrib/blob/master/test/test_swa.py#L337-L343

File: torchcontrib/optim/__init__.py (hunk 7 above)
Comment (izmailovpavel) [reply torchcontrib/optim/__init__.py]: done

File: test/common.py (hunk 9 above)
Comment (izmailovpavel) [reply test/common.py]: Done, removed most of the code from `common.py`

File: torchcontrib/optim/swa.py [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+            for p in group['params']:
+                param_state = self.state[p]
+                if 'swa_buffer' not in param_state:
+                    # If swa wasn't applied we don't swap params
+                    warnings.warn(
+                        "SWA wasn't applied to param {}; skipping it".format(p))
+                    continue
+                buf = param_state['swa_buffer']
+                tmp = torch.empty_like(p.data)
+                tmp.copy_(p.data)
```
Comment (apaszke) [inline torchcontrib/optim/swa.py]: `tmp = p.detach().clone()`
